### PR TITLE
Add new cop ChefModernize/ConditionalUsingTest

### PR DESCRIPTION
**Pull request by tas50**
Avoid shelling out when we can just use Ruby which is going to be faster.

Fixes https://github.com/chef/cookstyle/issues/543

Signed-off-by: Tim Smith <tsmith@chef.io>

### CODE REVIEW

File: spec/rubocop/cop/chef/modernize/conditional_using_test_spec.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require 'spec_helper'
+
+describe RuboCop::Cop::Chef::ChefModernize::ConditionalUsingTest, :config do
+  subject(:cop) { described_class.new(config) }
+
+  it 'registers an offense an only_if uses test -f' do
```
Comment (phiggins) [inline spec/rubocop/cop/chef/modernize/conditional_using_test_spec.rb:23]:
```suggestion
  it 'registers an offense when an only_if uses test -f' do
```

File: spec/rubocop/cop/chef/modernize/conditional_using_test_spec.rb [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    RUBY
+
+    expect_correction(<<~RUBY)
+      execute 'apt-get update' do
+        only_if { ::File.exist?('/sbin/apt') }
+      end
+    RUBY
+  end
+
+  it 'registers an offense a not_if uses test -f' do
```
Comment (phiggins) [inline spec/rubocop/cop/chef/modernize/conditional_using_test_spec.rb:38]:
```suggestion
  it 'registers an offense when a not_if uses test -f' do
```

File: spec/rubocop/cop/chef/modernize/conditional_using_test_spec.rb [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    RUBY
+
+    expect_correction(<<~RUBY)
+      execute 'yum update' do
+        not_if { ::File.exist?('/sbin/apt') }
+      end
+    RUBY
+  end
+
+  it "doesn't register an offense when test is used with a flag other and -e and -f" do
```
Comment (phiggins) [inline spec/rubocop/cop/chef/modernize/conditional_using_test_spec.rb:68]:
```suggestion
  it "doesn't register an offense when test is used with a flag other than -e or -f" do
```

File: spec/rubocop/cop/chef/modernize/conditional_using_test_spec.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        ^^^^^^^^^^^^^^^^^^^^^^^^^^ Use ::File.exist?('/foo/bar') instead of the slower 'test -f /foo/bar' which requires shelling out
+      end
+    RUBY
+
+    expect_correction(<<~RUBY)
+      execute 'yum update' do
+        not_if { ::File.exist?('/sbin/apt') }
+      end
+    RUBY
+  end
```
Comment (phiggins) [inline spec/rubocop/cop/chef/modernize/conditional_using_test_spec.rb:66]: I could see this being a tiny bit confusing if the user uses `test -e` and this suggests not using `test -f`.

Also this is the nitpickiest of nitpicks, but someone could want the behavior of `-f` that does more than test for existence.

Comment (tas50) [reply spec/rubocop/cop/chef/modernize/conditional_using_test_spec.rb:66]: Yeah -f is a bit different than File.exist?, but looking at how users were using it throughout supermarket cookbooks they were just looking for File.exist? functionality. Someone out there isn't going to like this and for that user they can disable it inline or entirely in a .rubocop.yml. For most people, I think it gives the appropriate direction.

**Pull request merged**